### PR TITLE
Download selic entries

### LINKED ISSUE #3: Create a method to download the SELIC entries for a given period
https://dadosabertos.bcb.gov.br/dataset/11-taxa-de-juros---selic/resource/b73edc07-bbac-430c-a2cb-b1639e605fa8

### PR DESCRIPTION
**Pull request by marileize**
Created a method to research the SELIC's entries for a given period. 

When merged, this will close #3

### CODE REVIEW

File: src/selic-entries.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export async function selicEntries(initialDate: string, finalDate: string) {
+    const url = new URL('https://api.bcb.gov.br/dados/serie/bcdata.sgs.11/dados?formato=json');
+    const params = new URLSearchParams;
+    params.append('dataInicial', initialDate);
+    params.append('dataFinal', finalDate);
```
Comment (adilsoncarvalho) [inline src/selic-entries.ts]: You can also assign an object to be taken as the parameters.

```suggestion
    const params = new URLSearchParams(
        dataInicial: initialDate,
        dataFinal: finalDate,
    );
```

File: src/selic-entries.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+export async function selicEntries(initialDate: string, finalDate: string) {
+    const url = new URL('https://api.bcb.gov.br/dados/serie/bcdata.sgs.11/dados?formato=json');
+    const params = new URLSearchParams;
```
Comment (adilsoncarvalho) [inline src/selic-entries.ts]: You missed parenthesis when instantiating a new instance of a class. I'm surprised that that worked.

```suggestion
    const params = new URLSearchParams();
```

File: src/selic-entries.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    params.append('dataInicial', initialDate);
+    params.append('dataFinal', finalDate);
+    url.search = params.toString();
+    
+    const response = await fetch(url.toString());
+    const selic_entries = await response.json();
+    return JSON.stringify(selic_entries);
```
Comment (adilsoncarvalho) [inline src/selic-entries.ts]: Why are you converting a data object into a string? One should be able to consume the outcome as an object

```json
[
  {
    "data":"08/03/2024",
    "valor":"0.041957"
  }
]
````

and not as a string

```typescript
'[{"data":"08/03/2024","valor":"0.041957"}]'
```

File: src/selic-entries.spec.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { selicEntries } from "./selic-entries";
+
+test("Testing selicEntries function", async () => {
+    expect(await selicEntries("08/03/2024","08/03/2024")).toBe('[{"data":"08/03/2024","valor":"0.041957"}]')
```
Comment (adilsoncarvalho) [inline src/selic-entries.spec.ts]: Why are we testing a string as the result of a method should return a json?

File: src/selic-entries.spec.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { selicEntries } from "./selic-entries";
+
+test("Testing selicEntries function", async () => {
```
Comment (adilsoncarvalho) [inline src/selic-entries.spec.ts]: The test name is redundant. Can you make it more clear to what it does test?

**Review by adilsoncarvalho — CHANGES_REQUESTED** [review]

**Comment by adilsoncarvalho** [comment]
I added a line to your description:

image

That line will [link this PR to the issue, and when this PR is merged, close that issue automatically](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

image

File: src/selic-entries.spec.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import { selicEntries } from "./selic-entries";
```
Comment (adilsoncarvalho) [inline src/selic-entries.spec.ts:1]: Hey, use single quotes whenever possible, ok?

```suggestion
// my fancy changes
import { selicEntries } from './selic-entries';
```

File: src/selic-entries.spec.ts (hunk 4 above)
Comment (adilsoncarvalho) [reply src/selic-entries.spec.ts]: Give it a read on how Jest do assertions: https://jestjs.io/docs/expect

There is one to compare objects that's neat.

File: src/selic-entries.spec.ts (hunk 5 above)
Comment (adilsoncarvalho) [reply src/selic-entries.spec.ts]: Try to namespace tests. Something like

```typescript
describe("selicEntries", () => {
  test("whatever tests", () => {
    // my funky test code is here ;)
  });
});
```

**Comment by marileize** [comment]
@adilsoncarvalho I want to add the API URL to a configuration file. Do you agree? If so, what is the recommended file?

**Comment by adilsoncarvalho** [comment]
Usually in node we don't do configuration files, but rather, get them from the environment. 

I need to find the link from the best practices when deploying services.

File: src/selic-entries.spec.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+describe("selicEntries", () => {
+    const desiredSelicEntrie = [
+        {
+            "data": '08/03/2024',
+            "valor": '0.041957',
+        }
+    ];
```
Comment (adilsoncarvalho) [inline src/selic-entries.spec.ts]: This shouldn't be here. When declared here, it'll become global to all the tests in this scope.

File: src/selic-entries.spec.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+    test("Successful request for a given period", async () => {
+        expect(await selicEntries("08/03/2024","08/03/2024")).toMatchObject(desiredSelicEntrie);
```
Comment (adilsoncarvalho) [inline src/selic-entries.spec.ts]: It's better to make the expectations more explicit.

```suggestion
        expect(await selicEntries("08/03/2024","08/03/2024")).toEqual({
            "data": '08/03/2024',
            "valor": '0.041957',
        });
```

File: src/selic-entries.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+export async function selicEntries(initialDate: string, finalDate: string) {
+    const url = new URL('https://api.bcb.gov.br/dados/serie/bcdata.sgs.11/dados?formato=json');
+    const params = new URLSearchParams([
+        ['dataInicial', initialDate],
+        ['dataFinal', finalDate],
+    ]);
+    url.search = params.toString();
```
Comment (adilsoncarvalho) [inline src/selic-entries.ts]: While this syntax you've chosen is ok, it doesn't look as good as using a hash. See the [examples](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams#examples) below:

```javascript
// Retrieve params via url.search, passed into constructor
const url = new URL("https://example.com?foo=1&bar=2");
const params1 = new URLSearchParams(url.search);

// Get the URLSearchParams object directly from a URL object
const params1a = url.searchParams;

// Pass in a string literal
const params2 = new URLSearchParams("foo=1&bar=2");
const params2a = new URLSearchParams("?foo=1&bar=2");

// Pass in a sequence of pairs
const params3 = new URLSearchParams([
  ["foo", "1"],
  ["bar", "2"],
]);

// Pass in a record
const params4 = new URLSearchParams({ foo: "1", bar: "2" });
```

---

```suggestion
    url.search = new URLSearchParams({
        dataInicial: initialDate,
        dataFinal: finalDate,
    }).toString();
```

File: src/selic-entries.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+    
+    const response = await fetch(url.toString());
+    const selic_entries = await response.json();
+    return selic_entries;
```
Comment (adilsoncarvalho) [inline src/selic-entries.ts]: Do we need this variable?

```suggestion
    return await response.json();
```

**Review by adilsoncarvalho — CHANGES_REQUESTED** [review]

**Comment by adilsoncarvalho** [comment]
- [ ] Read [How to Write a Git Commit Message](https://cbea.ms/git-commit/)

**Review by adilsoncarvalho — APPROVED** [review]
LGTM

**Pull request merged**